### DISCUSSION TITLE
I/O Access

### DISCUSSION THREAD

**Discussion by BenediktEnger** [simatic-ax/.discussions#12] [Ideas]
Why is SIMATIC AX not supporting I/O access? 
This is a strong limitation especially if you want to abstract the hardware layer to the rest of the logic.

Is it planned to change that in future?

**Comment by BeckerStS** [comment]
Hi @BenediktEnger 

SIMATIC AX ST does support I/O accessability in an **app** -lication environment. 
Have a look at this example (line12-14):
(https://github.com/simatic-ax/ae-trafficlight/blob/main/src/Configuration.st)


But we doubt you need direct IO access when developing **lib** -raries (with our TIAX usecase e.g.). 
Reusable libraries with fixed IO adresses feels a little bit odd to me. 
Can you pls. describe your usecase a little bit more in detail ?

Happy coding
🐱‍💻 BEEP, BOOP, BEEP, BEEP, BOOP 🐱‍🏍

**Reply by BenediktEnger** [reply]
For example: we want to have a wrapper around technology depending structures like drives. So we have a block, talking to the drive but with a generic interface. So it is easy to switch between drive vendors as long as the interface stays consistent. So the block is a connector between the block interface and the drive interface. 
But also, this block wants to send and receive data from the drive to exchange this interface. 

Because we try to create our program Independent of the technology used, we need to hide the hardware access behind generic interfaces. 
Of course, all those blocks I'm talking about are part of our current TIA Library (which is hard to handle with Git)

Edit: 
Of course I do not need the global addresses of the I/Os but the hardware ID is part of the block interface so the block can read and write using GET_IO/SET_IO (or others)
